### PR TITLE
Bug 1812813: oc debug node: create debug namespace with empty node-selector annotation

### PR DESCRIPTION
**Pull request by sallyom**
When a defaultNodeSelector is configured in the scheduler, `oc debug node` fails
if the node to debug does not have the matching labels.  To fix this, with
`oc debug node` create a temporary namespace for the debug pod that has an empty
`openshift.io/node-selector: ""` annotation.  Only create the debug ns if
 the user hasn't explicitly set the debug-pod namespace with
`--to-namespace` or `--namespace|-n`.

### CODE REVIEW

File: pkg/cli/debug/debug.go [hunk 1]
```diff
@@ -338,6 +341,11 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []s
 		return err
 	}
 
+	o.RBACClient, err = rbacv1client.NewForConfig(config)
```
Comment (deads2k) [inline pkg/cli/debug/debug.go:344]: if you can make a namespace, you don't need this binding.

File: pkg/cli/debug/debug.go [hunk 2]
```diff
@@ -375,11 +383,67 @@ func (o *DebugOptions) RunDebug() error {
...
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		// TODO fix this
+		//if debugNodeNS != nil {
+		//  o.Printer.PrintObj(debugNodeNS, o.Out)
+		//}
+		clusterRoleBinding, err = o.RBACClient.ClusterRoleBindings().Create(context.TODO(), o.newClusterRoleBinding(debugNodeNS.Name), metav1.CreateOptions{})
```
Comment (deads2k) [inline pkg/cli/debug/debug.go:408]: wrong scope

File: pkg/cli/debug/debug.go [hunk 3]
```diff
@@ -375,11 +383,67 @@ func (o *DebugOptions) RunDebug() error {
 	if err != nil {
 		klog.V(4).Infof("Unable to get exact template, but continuing with fallback: %v", err)
 	}
+	var debugNodeNS *corev1.Namespace
+	var clusterRoleBinding *rbacv1.ClusterRoleBinding
+	// create namespace only if debugging a node
+	if o.IsNode {
+		debugNodeNS, err = o.CoreClient.Namespaces().Create(context.TODO(), &corev1.Namespace{
```
Comment (deads2k) [inline pkg/cli/debug/debug.go:390]: only do this if the ns isn't explicitly set

File: pkg/cli/debug/debug.go (hunk 1 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: right, thanks - removed it!

File: pkg/cli/debug/debug.go (hunk 2 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: removed this

File: pkg/cli/debug/debug.go (hunk 3 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go:386]: right, fixed that, thanks again

File: pkg/cli/debug/debug.go [hunk 4]
```diff
@@ -375,11 +376,52 @@ func (o *DebugOptions) RunDebug() error {
 	if err != nil {
 		klog.V(4).Infof("Unable to get exact template, but continuing with fallback: %v", err)
 	}
+	var debugNodeNS *corev1.Namespace
+	generateName := names.SimpleNameGenerator.GenerateName("openshift-debug-node-")
+	// create namespace only if debugging a node
+	if o.IsNode && len(o.ToNamespace) == 0 && len(o.Namespace) == 0 {
+		if !o.Attach.Quiet {
+			fmt.Fprintf(o.Attach.ErrOut, "\nCreating debug namespace/%s ...\n", generateName)
```
Comment (soltysh) [inline pkg/cli/debug/debug.go:384]: `o.ErrOut`

Comment (soltysh) [reply pkg/cli/debug/debug.go:384]: Drop the newline at the beginning too

File: pkg/cli/debug/debug.go [hunk 5]
```diff
@@ -375,11 +376,52 @@ func (o *DebugOptions) RunDebug() error {
...
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	defer func() {
+		if debugNodeNS != nil {
+			if !o.Attach.Quiet {
+				fmt.Fprintf(o.Attach.ErrOut, "Removing debug namespace/%s ...\n", generateName)
```
Comment (soltysh) [inline pkg/cli/debug/debug.go:404]: `o.ErrOut`

File: pkg/cli/debug/debug.go [hunk 6]
```diff
@@ -375,11 +376,52 @@ func (o *DebugOptions) RunDebug() error {
...
+			}
+
+			if err := o.CoreClient.Namespaces().Delete(context.TODO(), debugNodeNS.Name, metav1.DeleteOptions{}); err != nil {
+				fmt.Printf("%v", err)
```
Comment (soltysh) [inline pkg/cli/debug/debug.go:408]: I'd print that to `o.ErrOut` just like above.

File: pkg/cli/debug/debug.go [hunk 7]
```diff
@@ -375,11 +376,52 @@ func (o *DebugOptions) RunDebug() error {
...
 		Spec:       template.Spec,
 	}
-	ns := infos[0].Namespace
+	ns := ""
+	if debugNodeNS != nil {
+		ns = debugNodeNS.Name
+	}
+	if len(ns) == 0 {
+		ns = infos[0].Namespace
+	}
```
Comment (soltysh) [inline pkg/cli/debug/debug.go:424]: The above will be simpler if you write it like this:
```go
ns := infos[0].Namespace
if debugNodeNS != nil {
    ns = debugNodeNS.Name
}
```

**Review by soltysh — APPROVED** [review]
A few nits
/approve

File: pkg/cli/debug/debug.go (hunk 4 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: updated, and also I changed the above`if len(o.Namespace) == 0`  to `if !o.ExplicitNamespace`  bc that was broken

File: pkg/cli/debug/debug.go (hunk 7 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: updated

File: pkg/cli/debug/debug.go (hunk 6 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: updated

File: pkg/cli/debug/debug.go (hunk 5 above)
Comment (sallyom) [reply pkg/cli/debug/debug.go]: updated

**Review by soltysh — APPROVED** [review]
/lgtm
/retest

**Comment by smarterclayton** [comment]
This behavior is really concerning.  Creating a new namespace is a big operation, and assumes a lot of privileges.  I don't think this is the appropriate fix for this problem, but instead, we should tell the user that they may not be able to schedule because they don't have permission to select.  A user can easily work around this problem by switching to another namespace, and creating a new namespace is a very, very dramatic operation.  It also violates a core assumption of debug - that the context you are running in doesn't change.

I would like to revert this and do a much simpler operation like warning the user when we don't schedule, not creating a namespace for them.

**Comment by sallyom** [comment]
> This behavior is really concerning. Creating a new namespace is a big operation, and assumes a lot of privileges. I don't think this is the appropriate fix for this problem, but instead, we should tell the user that they may not be able to schedule because they don't have permission to select. A user can easily work around this problem by switching to another namespace, and creating a new namespace is a very, very dramatic operation. It also violates a core assumption of debug - that the context you are running in doesn't change.
> 
> I would like to revert this and do a much simpler operation like warning the user when we don't schedule, not creating a namespace for them.

There was discussion in the bug report and on slack, and my initial feeling was the same :( 
https://bugzilla.redhat.com/show_bug.cgi?id=1812813#c35

**Comment by soltysh** [comment]
I agree with Clayton, that was my view when we initially put this together.

**Comment by sallyom** [comment]
I'll open a revert PR

**Pull request merged**